icepyx/core/variables.py

```diff
@@ -72,6 +74,14 @@ def __init__(
         elif self._vartype == "file":
             # DevGoal: check that the list or string are valid dir/files
             self.path = path
+        elif self._vartype == "nsidc-s3":
+            # Grab metadata from s3 path
+            template = ('s3://nsidc-cumulus-prod-protected/ATLAS/{product}/{version}/' 
```

Comment:
This template looks to be hardcoded to the non-gridded ATLAS datasets like ATL06. Other gridded products (e.g. ATL14, ATL16, ATL20) would have a different template. E.g. looking at https://search.earthdata.nasa.gov/search?ff=Available%20in%20Earthdata%20Cloud&fi=ATLAS&gdf=HDF&fst0=Cryosphere&lat=65.08299573518599&long=-25.69921875&zoom=5:

| Dataset | Sample path |
|--|--|
| ATL06 | s3://nsidc-cumulus-prod-protected/ATLAS/ATL06/006/2023/04/16/ATL06_20230416235213_04061911_006_02.h5 |
| ATL07 | s3://nsidc-cumulus-prod-protected/ATLAS/ATL07/005/2022/10/12/ATL07-02_20221012220720_03391701_005_01.h5 |
| ATL10 | s3://nsidc-cumulus-prod-protected/ATLAS/ATL10/005/2022/10/12/ATL10-01_20221012220720_03391701_005_01.h5 |
| ATL11 | s3://nsidc-cumulus-prod-protected/ATLAS/ATL11/005/2022/03/27/ATL11_006305_0315_005_03.h5 |
| ATL14 | s3://nsidc-cumulus-prod-protected/ATLAS/ATL14/002/2019/ATL14_IS_0314_100m_002_01.nc |
| ATL16 | s3://nsidc-cumulus-prod-protected/ATLAS/ATL16/004/2022/ATL16_20220722003637_04601601_004_01.h5 |
| ATL20 | s3://nsidc-cumulus-prod-protected/ATLAS/ATL20/003/2022/ATL20-01_20220901002201_10861601_003_01.h5 |

Would it be possible to generalize this code to both non-gridded and gridded products?

Reply:
Ah, thanks for catching this @weiji14! I'll work on a fix and let you know when I'm ready for review!

Reply:
A question for @JessicaS11 and @weiji14 -- What do we think of getting the version and product name from inside the file instead of parsing it from the filename? I've only checked a handful of products, but those fields seem to be available in top-level metadata in a consistent way. I've been trying to parse those things out of the filename, which is how I believe it is also done elsewhere in the module, but this limits the files icepyx can process to those named in a very specific way. If we grab product/version from inside the file we are able to process more files (ex. cloud icesat-2 files not in nsidc bucket, or local files that have had their name changed). Thoughts?

Reply:
The discussion summarized in https://github.com/icesat2py/icepyx/pull/438#issuecomment-1699753154 indicates our intention to move away from requiring the user provide the product as input (unless they are also feeding in a directory containing files from multiple products). This should address the template issues noted here.